Improve realm registration page usability

jackrzhang opened the pull request:
Still need to adjust the layout so that input fields are vertically aligned.

ping @rishig

smarx commented:
*Automated message from Dropbox CLA bot*

@jackrzhang, it looks like you've already signed the Dropbox CLA. Thanks!

hackerkid commented on templates/zerver/accounts_home.html:
```diff
@@ -55,7 +55,6 @@ <h1 class="get-started">{{ _("Sign up for Zulip") }}</h1>
                                     <input type="email" id="email" class="email" name="email" value="" required />
                                     <label for="email">Email</label>
                                     <div class="required"></div>
```
```html
 <div class="required"></div>
```
@jackrzhang Do we still need this line?

hackerkid commented on templates/zerver/register.html:
```diff
@@ -53,7 +53,7 @@
 
         <div class="input-box full-width">
             <div class="help-box">
-                {{ _('Zulip has full unicode support, so you can spell your name exactly how you want it to appear.') }}
```
> Spell your name however you like.


This does not convey any useful information to the use. We should probably just remove this message. The placeholder for this input already have a very good example. Also the box in which the message is shown is way too much big for this message.
![screenshot from 2017-06-24 00 24 56](https://user-images.githubusercontent.com/7190633/27496538-03e1cd74-5874-11e7-9a13-069f9d493822.png)

jackrzhang replied:
Perhaps `Full name or [full name in some other language].` would be more indicative. And yep - planning to change the size / layout of the help messages soon :)

hackerkid commented on templates/zerver/register.html:
```diff
@@ -40,7 +40,7 @@
             {% else %}
                 <input id="id_full_name" class="required" type="text" name="full_name"
                 value="{% if full_name %}{{ full_name }}{% elif form.full_name.value() %}{{ form.full_name.value() }}{% endif %}"
-                maxlength={{ MAX_NAME_LENGTH }} placeholder="Ex. Wḯ𝕝ë 𝔼. Ḉ𝕠yó𝕥é" required />
```
You should move this to the full name placeholder commit.

hackerkid commented on templates/zerver/register.html:
```diff
@@ -100,7 +100,7 @@
 
             <div class="help-box margin-top">
-                {{ _('This can be changed later on the settings page.') }}
+                {{ _('This can always be changed later. Keep it short!') }}
```
I feel it's kind of weird that we have an organization name field which encourage short names as well as another field called short name. (This field would be called Subdomain instead of Short name when subdomains are enabled). Is this change really necessary?

jackrzhang replied:
Yeah, @rishig and I had a discussion about this yesterday. Agreed that `short name` and `organization name` aren't too distinct as labels, though I think the note is still helpful. Also - planning on adding some logic so that the input value of the short name field is forced / automatically corrected to only lowercase letters, numbers, and dashes

rishig replied:
How about: "Shorter is better than longer. This can be changed later.", and then a tooltip that explains that this is used in notifications for users that enable it.

I'm not sure what to call the "short name" field in the non-subdomains case. I agree it is sort of awkward. Any ideas? One reason we need it is that the Organization Name can be in any language, but the Short Name has to be ascii (it is used in bot email addresses). 

Probably we should at minimum populate it automatically as people type in their organization name, if it's all ascii characters.

hackerkid commented on templates/zerver/register.html:
```diff
@@ -153,22 +153,16 @@
-              private message history. Zulip features that allow organization
-              administrators to take control of accounts or access private message
-              history for other users are disabled in community organizations.
+              Organization admins won't have access to take over other users'
```
We should probably make both the definitions sound similar. So we can probably change

`won't have access to` -> `won't be able to`
`see their private messages` -> `access their private messages`

Just a recommendation though.

jackrzhang replied:
agreed

hackerkid commented on templates/zerver/register.html:
```diff
@@ -153,22 +153,16 @@
-              administrators to take control of accounts or access private message
-              history for other users are only available to corporate organizations.
+              Organization admins will be able to take control of users'
+              accounts and access their private messages. Good for corporate
```
I am kind of against using **Good** here. It sounds like we recommend to use this mode for Corporations. I am not sure whether all the corporations need this features.

jackrzhang replied:
I believe that's the intention here. I'll leave this decision to @rishig

rishig replied:
Most corporations do need such features, for better or worse. Often for compliance or other legal reasons. This is actually the default for most products; e.g. the admin of a slack team can take control of user accounts and access their private messages (probably not on their free tier, but I expect at some tier).

I agree this section needs some thought; let's leave it as is for now, and then figure out the exact thing to do when Tim gets back.

jackrzhang commented:
Screenshot: 

![image](https://user-images.githubusercontent.com/12771126/27503181-d713e2c2-582e-11e7-9a54-45a75c106d3e.png)

Haven't touched any css yet; layout to be changed.

rishig commented on templates/zerver/register.html:
```diff
@@ -40,7 +40,7 @@
                 <input id="id_full_name" class="required" type="text" name="full_name"
                 value="{% if full_name %}{{ full_name }}{% elif form.full_name.value() %}{{ form.full_name.value() }}{% endif %}"
-                maxlength={{ MAX_NAME_LENGTH }} required />
+                maxlength={{ MAX_NAME_LENGTH }} placeholder="Full name or 名前" required />
```
This should be squashed with the previous commit.

rishig commented on templates/zerver/register.html:
```diff
@@ -81,7 +81,7 @@
             <div class="inline-block relative">
                 <input id="id_team_name" class="required" type="text"
-                    placeholder="Acme"
+                    placeholder="Äcmé"
```
Changing my mind back to A without the umlaut (I think it looks like we're trying too hard with both :)).

rishig commented on templates/zerver/register.html:
```diff
@@ -109,7 +109,7 @@
                 <div class="inline-block relative">
                     <input id="id_team_subdomain" class="required" type="text"
-                        placeholder="acme"
+                        placeholder="Ex. acme-sales"
```
Lets leave this as `acme`

rishig replied:
This gets changed again later, I think

rishig commented on templates/zerver/register.html:
```diff
@@ -124,10 +124,9 @@
             <div class="help-box margin-top">
+                {{ _("Letters, numbers, and dashes only.") }}
                 {% if realms_have_subdomains %}
-                    {{ _("The address you'll use to sign in to your organization.") }}
```
I think we can keep this if-else structure. a-z, 0-9 etc in the non-subdomains case, and I think we don't have to say it in the subdomains case. 

"the address you'll use to sign in to your organization" does feel more clear than than the shorter version.

rishig commented on templates/zerver/register.html:
```diff
@@ -206,6 +206,10 @@
 
 $("#timezone").val(moment.tz.guess());
+
+$('#id_team_subdomain').on('input', function (e) {
```
I think we can do something slightly nicer, which is 
* replace upper case with lower case
* replace space and _ with -
* if the user types anything else (or starts with a -), show an error message

jackrzhang commented:
fullscreen:
![image](https://user-images.githubusercontent.com/12771126/27560554-7b591e1e-5a79-11e7-8bf8-c6a6d8114565.png)

mobile:
![image](https://user-images.githubusercontent.com/12771126/27560621-bdaf57ce-5a79-11e7-9f91-c7eba3d01239.png)

ping @rishig

rishig commented:
(comments given off-site, no review needed yet)

jackrzhang commented:
Removed all help boxes and (temporarily) the radio buttons for org type. The form is definitely looking a lot simpler 👌 

![image](https://user-images.githubusercontent.com/12771126/27562834-51ac8fa6-5a83-11e7-8d3a-8f748920e69e.png)

![image](https://user-images.githubusercontent.com/12771126/27562915-ae975868-5a83-11e7-83d9-990a4a4e4c66.png)

jackrzhang commented:
@rishig 

![image](https://user-images.githubusercontent.com/12771126/27887512-f01f598e-6195-11e7-8792-45205978e724.png)

jackrzhang commented:
Subdomain input field:

@rishig 
![image](https://user-images.githubusercontent.com/12771126/27888869-5d026bac-619d-11e7-9e25-2b0294466908.png)

rishig commented on templates/zerver/register.html:
```diff
@@ -49,7 +48,8 @@
                 {% endif %}
             {% endif %}
-            <img class="valid" src="/static/images/checkbox-valid.svg" alt="{{ _('Valid') }}" />
+            <label for="id_full_name" class="inline-block label-title">{{ _('Full name') }}</label>
```
does this belong to a future commit?

rishig commented on templates/zerver/register.html:
```diff
@@ -210,9 +210,9 @@
 
 <script type="text/javascript">
-if ($('#id_email:visible').length) {
+if (!$('#id_email').attr('disabled')) {
```
should also work when `#id_email` doesn't exist

rishig commented on templates/zerver/register.html:
```diff
@@ -210,9 +210,9 @@
+if (!$('#id_email').attr('disabled')) {
     common.autofocus('#id_email');
-} else if ($('#id_full_name:visible').length) {
+} else if (!$('#id_full_name').attr('disabled')) {
```
`#id_full_name` will never be disabled, only id_email.

rishig commented on templates/zerver/register.html:
```diff
@@ -40,8 +40,7 @@
                 <input id="id_full_name" class="required" type="text" name="full_name"
                 value="{% if full_name %}{{ full_name }}{% elif form.full_name.value() %}{{ form.full_name.value() }}{% endif %}"
-                maxlength={{ MAX_NAME_LENGTH }} required />
-                <label for="id_full_name" class="inline-block label-title">{{ _('Full name') }}</label>
```
ah, I think this was added in the first commit?

rishig commented on templates/zerver/register.html:
```diff
@@ -40,8 +40,7 @@
                 value="{% if full_name %}{{ full_name }}{% elif form.full_name.value() %}{{ form.full_name.value() }}{% endif %}"
-                maxlength={{ MAX_NAME_LENGTH }} required />
-                <label for="id_full_name" class="inline-block label-title">{{ _('Full name') }}</label>
+                maxlength={{ MAX_NAME_LENGTH }} placeholder="{% trans %}Full name{% endtrans %} or 名前" required />
```
The translation here is a bit tricky, since "or" also needs to be translated, and for the japanese translation we wouldn't want "名前 or 名前". Maybe put the whole thing into the `{% trans %}`, and the translators can figure out what to do from context?

jackrzhang replied:
Ah, that's true. Sounds good

rishig commented on templates/zerver/register.html:
```diff
@@ -131,10 +124,9 @@
                 {% endif %}
             </div>
             <div class="help-box margin-top">
+                {{ _("Letters, numbers, and dashes only.") }}
```
This is changed again later, right? we can just change it in this commit to "a-z, 0-9, .."

rishig commented on templates/zerver/register.html:
```diff
@@ -155,22 +147,16 @@
-              private message history. Zulip features that allow organization
-              administrators to take control of accounts or access private message
-              history for other users are only available to corporate organizations.
+              Organization admins will be able to take control of users'
```
I would just move the commit that removes the org type entirely to come before this commit.

rishig commented on tools/linter_lib/custom_check.py:
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
```
This change looks like it belongs to another commit?

jackrzhang replied:
It's needed for the characters "名前" to pass the linter. Unless there's a more direct way to add an exception for those characters?

jackrzhang replied:
Get this error without it:
```
SyntaxError: Non-ASCII character '\xce' in file /srv/zulip/tools/linter_lib/custom_check.py on line 350, but no encoding declared; see http://www.python.org/peps/pep-0263.html for details
```

jackrzhang commented:
Squashed some commits and organized the history a bit more.

rishig commented:
cool, thanks for all your work on this @jackrzhang! 
I did some additional cleanup and pushed a version to https://github.com/rishig/zulip/tree/review-5532. 
The commits before "portico: Rearrange registration form into centered column." I think are in good shape, but that commit is hard to grok, mostly due to all the reindenting. 

Could you 
* Split that commit up into smaller commits if needed (there still seem to be a couple of things going on there?)
* Do all the real work without reindenting (the first commit can add `templates/zerver/register.html` into IGNORE_FILES of tools/check-templates), and then reindent at the end (using pretty_print.py or something; hopefully you aren't doing it by hand).

Thanks!

timabbott commented:
I think the conflicts are with the new password change code.  Would love to start merging this soon.

rishig commented:
@jackrzhang, I should have mentioned; one thing that's helpful is to force push https://github.com/rishig/zulip/tree/review-5532 onto this right away, so part of it can start being merged, even as you work on the later commits.

jackrzhang commented:
@rishig I removed `register.html` from IGNORE_FILES of tools/check_templates  and rewrote some of the commits in a cleaner fashion. That last commit, `registration: Divide form into user and org sections.`, just indents and wraps the form input boxes with section tags. The diff looks a bit messy though :/

timabbott commented:
@jackrzhang what's the story with the 00 casper tests in the 3rd commit?  It seems like you're removing the check that the user ended up logged in at the end of registering a realm, which I don't understand (there also seem to be merge conflicts with a fix I did for that test flaking).

jackrzhang commented:
Those tests were flaking when I initially made the 3rd commit (Casper's error message was pretty unhelpful :( ), so I removed them. I've rebased the branch and they pass consistently now, so I've reverted all of the changes I made to the file.

rishig commented on templates/zerver/register.html:
```diff
@@ -105,10 +105,13 @@
                             value="{% if form.realm_subdomain.value() %}{{ form.realm_subdomain.value() }}{% endif %}"
                             name="realm_subdomain" maxlength={{ MAX_REALM_SUBDOMAIN_LENGTH }} required />
+                        {% if realms_have_subdomains %}
+                        <p class="realm_subdomain_label">.host.com</p>
```
This should be `.{{ external_host }}`

jackrzhang replied:
Updated.

jackrzhang commented:
* text align error messages / help messages to the left
* look into zxcvbn password bar (update on page reload)
* Client-side error messages should not shift the height of the form

rishig commented:
Also: last two commits (spacing from host.com, and split last commit into "real" changes and indentation.)
Thanks @jackrzhang!

jackrzhang commented:
@rishig updated

rishig commented:
lgtm, thanks @jackrzhang!
Ready for review @timabbott. (I did some manual testing and cursory review of the code.)

As follow-up we could figure out slightly shorter wording for "The address you'll use to sign in to your organization." so that it fits in one line, but we can deal with that later.

timabbott commented:
@jackrzhang this looks great!  Can you rebase onto master to resolve the merge conflicts in `success-http-headers`?  I think we'll need to just have the tests fail in Travis to get the correct values.

Some follow-up projects we should think about:
* Do something to replace the `success-http-headers` system.  At the very least, we want something that one can run locally (could try just extracting from `tools/travis/production-helper`) and potentially we also want to replace it with a more comprehensive/value focused test.  Right now the only true errors it catches are "Zulip server doesn't start and server logged-out pages", and it has these false positive issues.  Ideally, we'd run a simple Casper test against the new server instead.  
* Add an API endpoint that lets you check if a handle is available via an AJAX request so you don't have to hit submit to find out the value is invalid.

jackrzhang commented:
@timabbott This is good to merge.

timabbott commented:
Merged, thanks @jackrzhang!

We should be sure to do a full test of this when it gets onto chat.zulip.org; I'll do that now.
